### PR TITLE
Modify FLAME output and add post-matching functions

### PR DESCRIPTION
**Pull request by thowell332**
Not ready to merge yet.

### CODE REVIEW

File: dame_flame/post_matching.py
Comment (nehargupta) [inline dame_flame/post_matching.py:24]: I think this #%% gets auto added if you're developing in Jupyter or something? Can you remove them somehow bc we want to avoid inconsistent formatting

File: dame_flame/post_matching.py
Comment (nehargupta) [inline dame_flame/post_matching.py:40]: nitpicky thing but also for consistent formatting, can we stick with  comments being hash + space + start of comment? I tried (and sometimes was also lazy) to stick with Google's Python Style Guide, so that's where stylistic conventions I'm recommending you change come from:

https://google.github.io/styleguide/pyguide.html

But these are great comments content wise so if you did a similar thing in the FLAME algorithm that would be great!

File: dame_flame/flame_algorithm.py
Comment (nehargupta) [inline dame_flame/flame_algorithm.py:98]: Clean up formatting yes, but can you also put a one line comment in, detailing what the point of this loop is in plain language? ie I have stuff like "Now remove the matched units", "here we begin the iterative dropping procedure"...

File: dame_flame/grouped_mr.py
Comment (nehargupta) [inline dame_flame/grouped_mr.py:53]: Comment detailing what this will store, like the above lines

File: dame_flame/grouped_mr.py
Comment (nehargupta) [inline dame_flame/grouped_mr.py:84]: Need to update the header of this function to reflect the new parameter being returned from the function

File: dame_flame/flame_algorithm.py
Comment (nehargupta) [inline dame_flame/flame_algorithm.py:306]: remove # == formatting

File: dame_flame/flame_algorithm.py
Comment (nehargupta) [inline dame_flame/flame_algorithm.py:201]: remove # = formatting and commented out code unless accompanied by a todo. Comment detailing the point of this code block

File: dame_flame/post_matching.py
Comment (nehargupta) [inline dame_flame/post_matching.py:35]: Add what is being returned by the function and the data type as well please!

File: dame_flame/post_matching.py
Comment (nehargupta) [inline dame_flame/post_matching.py:44]: Can you not use 'i' and 'j' in loops like this unless it's really needed? 

So maybe for group_number in MGs or something like that, whatever is more descriptive.

File: dame_flame/post_matching.py
Comment (nehargupta) [reply dame_flame/post_matching.py:35]: Same for all functions here

File: dame_flame/post_matching.py
Comment (nehargupta) [inline dame_flame/post_matching.py:86]: is this how the multi-line formatting should look?

File: dame_flame/post_matching.py
Comment (nehargupta) [inline dame_flame/post_matching.py:46]: for this function and all the below ones, I try to follow the rule of making sure to explain any nested loops, either in a comment above the outer or in the inner one

something like "iterate through blah and for each one, find the corresponding blah in the blah"

File: dame_flame/post_matching.py
Comment (nehargupta) [inline dame_flame/post_matching.py:31]: Since you updated what this looks like, could you update the documentation that describes this? I imagine its definitely described in the README but possibly also in DAME_FLAME.py

File: dame_flame/post_matching.py
Comment (nehargupta) [inline dame_flame/post_matching.py:141]: I think you did use a separate file for this and only ran it separately, but we need to integrate it in the package. I don't think having code outside of a function will error if you leave this here

**Comment by thowell332** [comment]
TO DO: update README, apply changes to DAME algorithm, update documentation within dame scripts

I'm holding off on these to-do actions for the moment - once my code for FLAME is approved I will apply the changes to DAME and update all relevant documentation!

**Comment by thowell332** [comment]
NOTE: I did not delete the original post-matching functions or either of the query scripts in case you want to compare the new functions to the old ones.

File: dame_flame/grouped_mr.py
Comment (nehargupta) [inline dame_flame/grouped_mr.py:61]: Found the issue you were encountering on sample.csv (I'm pretty sure, but do double check me on this). 

So that dataset first dropped x4 and found a group of units 0,1. Cool. 
Then on the second iteration, it's here and considering dropping x3. When we look at who we can match on x3&x4, then units 0&1 match again and form a group again. So, 0,1 is units_in_g. But, the line below that is where it checks to see if any of the units in the group [0,1] already have groups formed, and the if statement below that is where it makes sure it only adds to return_groups (the listing with the "*"s in the event that a new MMG is being created. 

So, you'll want to move the line where you add to all_units_in_g after the `if len(newly_matched) !=0`. The presence of that if-statement is why return_groups only had one entry for 0&1 but your units_in_g had multiple entries of those. 

(I suggest to debug this by sticking a print statement in for units_in_g and newly_matched right when they're first created)

File: dame_flame/grouped_mr.py
Comment (nehargupta) [inline dame_flame/grouped_mr.py:55]: # This list stores all unique matched groups created by this function....
+ " where each matched group is represented as a list of units" or something indicating what the entries of the list are. This is nitpicky thing, I understood it this time with this comment but maybe someone less with it than me wouldn't? idk

File: README.md
Comment (nehargupta) [inline README.md:80]: "the number of times each unit was matched" could be confusing bc does that mean the number of units I was matched to or the number of groups I'm in?

So maybe, "the number of main matched groups a unit is part of" or something? 

Can you detail the second element a little more? I think I said something about this below as well, in the grouped_mr.py comments. So it's a list of lists, where each list contains the unit ids belonging to a main matched group? Or is it any group beyond just main matched groups -- do clarify this.

Comment (nehargupta) [reply README.md:80]: Also I'm realizing that hopefully this isn't confusing: "The list 'result' will have additional values based on additional optional parameters, detailed in additional documentation below."

Would you be able to check in the documentation for where we mention the additional optional parameters and make sure we don't call them the "1st" and "2nd" or something like that, because now the ordinality would be changed.

I guess if we wanted we could make your list of matched group ids also an optional parameter to return and even store/compute. If we did that though, we would need two versions of the find_mmg_of_unit function, etc, so it's maybe not worth the hassle? I recall you saying you thought adding these new data structs wasn't really a change to the time complexity in your speed trials, but if it is and it's worth it, go ahead with it I guess?

File: README.md
Comment (nehargupta) [inline README.md:85]: I guess you must have checked with the team to confirm you like the new naming of these functions? I'm fine with whatever personally

File: README.md
Comment (nehargupta) [inline README.md:113]: Could combine these into one codeblock/comment block if you wanted? Up to how you think it looks better, just pointing out that you have the option :)

File: dame_flame/dame_algorithm.py
Comment (nehargupta) [inline dame_flame/dame_algorithm.py:113]: This file looks great good job with the fast changes on it!

File: dame_flame/grouped_mr.py
Comment (thowell332) [reply dame_flame/grouped_mr.py:61]: That is totally the issue. Thanks for catching this!

File: README.md
Comment (thowell332) [reply README.md:85]: Oh I actually forgot to ask about the names - I will send a slack message now

File: README.md
Comment (thowell332) [reply README.md:80]: I do not see any mention of the ordinality of the optional parameters, so I think we're good on that front!

As for making the second element of the list 'result' optional, that is a very interesting idea. I think it would be worth it if we discover that computing MGs significantly affects the algorithm's speed on big datasets. If we do make it optional, we already have your mmg_of_unit and te_of_unit functions which are equipped to recover that information without the additional return. We would just need to update the mmg function to work when repeats = False. Also I think it would be a good idea to merge our mmg and te functions into two functions (as opposed to four) which can determine which method to use on their own. Other than that, the ATE function would need to be modified to be able to recover a complete list of MGs post-matching, but I do believe that this is possible (albeit slow).

File: README.md
Comment (thowell332) [reply README.md:113]: Do you mean something like this?

```Python
ate = dame_flame.DAME_FLAME.ATE(return_df=result, input_data=df)
att = dame_flame.DAME_FLAME.ATT(return_df=result, input_data=df)
print(ate, att)
#> 2.0  2.0
```

Comment (nehargupta) [reply README.md:113]: I meant that you go:
- text description of cate
- code block of cate
- text description of att/ate
- code block of att/ate

so if you wanted you could go
- text description of cate/att/ate
- code block of all 3

or 6 alternating text + code blocks and separate att/ate...?

up to you whatever you think looks best aesthetically, I'm fine with whatever

File: README.md
Comment (nehargupta) [reply README.md:80]: RE: "I think it would be worth it if we discover that computing MGs significantly affects the algorithm's speed on big datasets"

If I remember correctly, you said you didn't think it made a speed decrease -- do you remember what the size you ran it on for that was?

Comment (thowell332) [reply README.md]: For 5000 units with 20 covariates the modified version was about 2 seconds faster and took about 87 seconds. For 10000 units with 30 covariates, the modified version was about 3 seconds slower and took about 175 seconds. I haven't tested with any larger datasets though.

**Review by nehargupta — COMMENTED** [review]
Code looks good. I am going to download and run it a bit, and then will approve shortly. I think I am the one who can merge them in, or possibly you can, not sure. Do you want all the data files? I don't have preference, just checking with you first. Also, let me know if you have preference on how the merge goes in...it looks like I can squash your 27 commits into 1 or keep them all, I'm fine with either way

**Review by nehargupta — APPROVED** [review]
Ok! I ran the code and compared it to some test files I have and the output looked good/unchanged so basically it is what we think it should be. Approving to push! Let me know if I should merge or you. Not sure if you have privileges to do so, but if you do, then I think you doing it would be preferred for keeping your name on it

**Pull request merged**
